File: skills/optimization/reference/src/line-search.ts
/**
 * Line search algorithms for gradient-based optimization.
 *
 * Given a position x, direction d, and step size alpha, find an alpha that
 * satisfies sufficient decrease conditions. Two strategies:
 *
 * 1. Backtracking (Armijo): simple, robust, good default for gradient descent
 * 2. Strong Wolfe: required for BFGS (ensures positive-definite Hessian updates)
 *
 * @node line-search
 * @depends-on vec-ops
 * @contract line-search.test.ts
 * @hint wolfe-params: c1=1e-4, c2=0.9 matches scipy. Optim.jl's HagerZhang uses
 *       different parameters internally. Our choice: standard Wolfe from Nocedal & Wright.
 * @hint return: Returns { alpha, fNew, gNew, calls } so the caller can avoid
 *       redundant function/gradient evaluations.
 */

import { dot, addScaled } from "./vec-ops";

export interface LineSearchResult {
  /** Step size found. */
  alpha: number;
  /** Function value at x + alpha * d. */
  fNew: number;
  /** Gradient at x + alpha * d (if computed, else null). */
  gNew: number[] | null;
  /** Number of function evaluations used. */
  functionCalls: number;
  /** Number of gradient evaluations used. */
  gradientCalls: number;
  /** Whether the line search succeeded. */
  success: boolean;
}

/**
 * Backtracking line search with Armijo (sufficient decrease) condition.
 *
 * Finds alpha such that: f(x + alpha*d) <= f(x) + c1*alpha*g'*d
 *
 * Starts with alpha=initialAlpha and multiplies by rho until the condition
 * is satisfied or maxIter is reached.
 *
 * @provenance Nocedal & Wright, Numerical Optimization, Algorithm 3.1
 * @provenance c1=1e-4 matches scipy.optimize.line_search and Optim.jl
 * @provenance rho=0.5 (halving) is the simplest; scipy uses cubic interpolation
 */
export function backtrackingLineSearch(
  f: (x: number[]) => number,
  x: number[],
  d: number[],
  fx: number,
  gx: number[],
  options?: {
    initialAlpha?: number;
    c1?: number;
    rho?: number;
    maxIter?: number;
  },
): LineSearchResult {
  const initialAlpha = options?.initialAlpha ?? 1.0;
  const c1 = options?.c1 ?? 1e-4;
  const rho = options?.rho ?? 0.5;
  const maxIter = options?.maxIter ?? 20;

  const dg = dot(gx, d); // directional derivative g'*d (should be negative)
  let alpha = initialAlpha;
  let functionCalls = 0;

  for (let i = 0; i < maxIter; i++) {
    const xNew = addScaled(x, d, alpha);
    const fNew = f(xNew);
    functionCalls++;

    // Armijo condition: f(x + alpha*d) <= f(x) + c1*alpha*(g'*d)
    if (fNew <= fx + c1 * alpha * dg) {
      return {
        alpha,
        fNew,
        gNew: null,
        functionCalls,
        gradientCalls: 0,
        success: true,
      };
    }

    alpha *= rho;
  }

  // Failed to find a step satisfying Armijo condition
  return {
    alpha,
    fNew: f(addScaled(x, d, alpha)),
    gNew: null,
    functionCalls: functionCalls + 1,
    gradientCalls: 0,
    success: false,
  };
}

/**
 * Strong Wolfe line search.
 *
 * Finds alpha satisfying both:
 *   1. Sufficient decrease (Armijo): f(x + alpha*d) <= f(x) + c1*alpha*g'*d
 *   2. Curvature condition: |g(x + alpha*d)'*d| <= c2*|g(x)'*d|
 *
 * Uses the bracket-and-zoom approach from Nocedal & Wright Chapter 3.
 * Required for BFGS to maintain positive-definite Hessian approximation.
 *
 * @provenance Nocedal & Wright, Numerical Optimization, Algorithm 3.5 + 3.6
 * @provenance c1=1e-4, c2=0.9 matches scipy.optimize.line_search defaults
 * @provenance Optim.jl uses HagerZhang by default (different algorithm, similar guarantees)
 */
export function wolfeLineSearch(
  f: (x: number[]) => number,
  grad: (x: number[]) => number[],
  x: number[],
  d: number[],
  fx: number,
  gx: number[],
  options?: {
    c1?: number;
    c2?: number;
    alphaMax?: number;
    maxIter?: number;
  },
): LineSearchResult {
  const c1 = options?.c1 ?? 1e-4;
  const c2 = options?.c2 ?? 0.9;
  const alphaMax = options?.alphaMax ?? 1e6;
  const maxIter = options?.maxIter ?? 25;

  const dg0 = dot(gx, d); // initial directional derivative
  let functionCalls = 0;
  let gradientCalls = 0;

  let alphaPrev = 0;
  let fPrev = fx;
  let alpha = 1.0;

  for (let i = 0; i < maxIter; i++) {
    const xNew = addScaled(x, d, alpha);
    const fNew = f(xNew);
    functionCalls++;

    // Check Armijo condition or if function increased from previous
    if (fNew > fx + c1 * alpha * dg0 || (i > 0 && fNew >= fPrev)) {
      return zoom(f, grad, x, d, fx, dg0, c1, c2, alphaPrev, alpha, fPrev, fNew,
        functionCalls, gradientCalls);
    }

    const gNew = grad(xNew);
    gradientCalls++;
    const dgNew = dot(gNew, d);

    // Check curvature condition
    if (Math.abs(dgNew) <= c2 * Math.abs(dg0)) {
      return {
        alpha, fNew, gNew, functionCalls, gradientCalls, success: true,
      };
    }

    // If directional derivative is positive, we've bracketed
    if (dgNew >= 0) {
      return zoom(f, grad, x, d, fx, dg0, c1, c2, alpha, alphaPrev, fNew, fPrev,
        functionCalls, gradientCalls);
    }

    alphaPrev = alpha;
    fPrev = fNew;
    alpha = Math.min(2 * alpha, alphaMax);
  }

  // Failed — return best we have
  const xFinal = addScaled(x, d, alpha);
  return {
    alpha,
    fNew: f(xFinal),
    gNew: grad(xFinal),
    functionCalls: functionCalls + 1,
    gradientCalls: gradientCalls + 1,
    success: false,
  };
}

/**
 * Zoom phase of the Wolfe line search.
 * Narrows the bracket [alphaLo, alphaHi] to find a point satisfying both conditions.
 *
 * @provenance Nocedal & Wright, Algorithm 3.6
 */
function zoom(
  f: (x: number[]) => number,
  grad: (x: number[]) => number[],
  x: number[],
  d: number[],
  fx: number,
  dg0: number,
  c1: number,
  c2: number,
  alphaLo: number,
  alphaHi: number,
  fLo: number,
  fHi: number,
  functionCalls: number,
  gradientCalls: number,
): LineSearchResult {
  const maxZoomIter = 20;

  for (let j = 0; j < maxZoomIter; j++) {
    // Bisect the interval
    const alpha = (alphaLo + alphaHi) / 2;
    const xNew = addScaled(x, d, alpha);
    const fNew = f(xNew);
    functionCalls++;

    if (fNew > fx + c1 * alpha * dg0 || fNew >= fLo) {
      alphaHi = alpha;
      fHi = fNew;
    } else {
      const gNew = grad(xNew);
      gradientCalls++;
      const dgNew = dot(gNew, d);

      if (Math.abs(dgNew) <= c2 * Math.abs(dg0)) {
        return {
          alpha, fNew, gNew, functionCalls, gradientCalls, success: true,
        };
      }

      if (dgNew * (alphaHi - alphaLo) >= 0) {
        alphaHi = alphaLo;
        fHi = fLo;
      }

      alphaLo = alpha;
      fLo = fNew;
    }

    // Bracket too small
    if (Math.abs(alphaHi - alphaLo) < 1e-14) {
      break;
    }
  }

  // Return best we found (alphaLo is usually the better end)
  const xFinal = addScaled(x, d, alphaLo);
  return {
    alpha: alphaLo,
    fNew: f(xFinal),
    gNew: grad(xFinal),
    functionCalls: functionCalls + 1,
    gradientCalls: gradientCalls + 1,
    success: false,
  };
}

File: research/experiments/sweep/runs/opt-medium-python-opus45/src/line_search.py
"""
Line search algorithms: backtracking (Armijo) and Strong Wolfe.

Provenance: Nocedal & Wright, Numerical Optimization, Algorithms 3.1, 3.5, 3.6
"""

from dataclasses import dataclass
from typing import Callable, Optional

from vec_ops import dot, add_scaled


@dataclass
class LineSearchResult:
    alpha: float
    f_new: float
    g_new: Optional[list[float]]
    function_calls: int
    gradient_calls: int
    success: bool


def backtracking_line_search(
    f: Callable[[list[float]], float],
    x: list[float],
    d: list[float],
    fx: float,
    gx: list[float],
    initial_alpha: float = 1.0,
    c1: float = 1e-4,
    rho: float = 0.5,
    max_iter: int = 20,
) -> LineSearchResult:
    """
    Backtracking line search with Armijo (sufficient decrease) condition.
    Finds alpha such that: f(x + alpha*d) <= f(x) + c1*alpha*g'*d
    """
    dg = dot(gx, d)
    alpha = initial_alpha
    function_calls = 0

    for _ in range(max_iter):
        x_new = add_scaled(x, d, alpha)
        f_new = f(x_new)
        function_calls += 1

        if f_new <= fx + c1 * alpha * dg:
            return LineSearchResult(
                alpha=alpha,
                f_new=f_new,
                g_new=None,
                function_calls=function_calls,
                gradient_calls=0,
                success=True,
            )

        alpha *= rho

    # Failed to find a step satisfying Armijo condition
    x_final = add_scaled(x, d, alpha)
    return LineSearchResult(
        alpha=alpha,
        f_new=f(x_final),
        g_new=None,
        function_calls=function_calls + 1,
        gradient_calls=0,
        success=False,
    )


def wolfe_line_search(
    f: Callable[[list[float]], float],
    grad: Callable[[list[float]], list[float]],
    x: list[float],
    d: list[float],
    fx: float,
    gx: list[float],
    c1: float = 1e-4,
    c2: float = 0.9,
    alpha_max: float = 1e6,
    max_iter: int = 25,
) -> LineSearchResult:
    """
    Strong Wolfe line search (bracket-and-zoom).
    Finds alpha satisfying both Armijo and curvature conditions.
    """
    dg0 = dot(gx, d)
    function_calls = 0
    gradient_calls = 0

    alpha_prev = 0.0
    f_prev = fx
    alpha = 1.0

    for i in range(max_iter):
        x_new = add_scaled(x, d, alpha)
        f_new = f(x_new)
        function_calls += 1

        # Check Armijo condition or if function increased from previous
        if f_new > fx + c1 * alpha * dg0 or (i > 0 and f_new >= f_prev):
            return _zoom(
                f, grad, x, d, fx, dg0, c1, c2,
                alpha_prev, alpha, f_prev, f_new,
                function_calls, gradient_calls,
            )

        g_new = grad(x_new)
        gradient_calls += 1
        dg_new = dot(g_new, d)

        # Check curvature condition
        if abs(dg_new) <= c2 * abs(dg0):
            return LineSearchResult(
                alpha=alpha,
                f_new=f_new,
                g_new=g_new,
                function_calls=function_calls,
                gradient_calls=gradient_calls,
                success=True,
            )

        # If directional derivative is positive, we've bracketed
        if dg_new >= 0:
            return _zoom(
                f, grad, x, d, fx, dg0, c1, c2,
                alpha, alpha_prev, f_new, f_prev,
                function_calls, gradient_calls,
            )

        alpha_prev = alpha
        f_prev = f_new
        alpha = min(2 * alpha, alpha_max)

    # Failed -- return best we have
    x_final = add_scaled(x, d, alpha)
    return LineSearchResult(
        alpha=alpha,
        f_new=f(x_final),
        g_new=grad(x_final),
        function_calls=function_calls + 1,
        gradient_calls=gradient_calls + 1,
        success=False,
    )


def _zoom(
    f: Callable[[list[float]], float],
    grad: Callable[[list[float]], list[float]],
    x: list[float],
    d: list[float],
    fx: float,
    dg0: float,
    c1: float,
    c2: float,
    alpha_lo: float,
    alpha_hi: float,
    f_lo: float,
    f_hi: float,
    function_calls: int,
    gradient_calls: int,
) -> LineSearchResult:
    """
    Zoom phase of the Wolfe line search.
    Narrows the bracket [alpha_lo, alpha_hi] to find a point satisfying both conditions.
    """
    max_zoom_iter = 20

    for _ in range(max_zoom_iter):
        alpha = (alpha_lo + alpha_hi) / 2
        x_new = add_scaled(x, d, alpha)
        f_new = f(x_new)
        function_calls += 1

        if f_new > fx + c1 * alpha * dg0 or f_new >= f_lo:
            alpha_hi = alpha
            f_hi = f_new
        else:
            g_new = grad(x_new)
            gradient_calls += 1
            dg_new = dot(g_new, d)

            if abs(dg_new) <= c2 * abs(dg0):
                return LineSearchResult(
                    alpha=alpha,
                    f_new=f_new,
                    g_new=g_new,
                    function_calls=function_calls,
                    gradient_calls=gradient_calls,
                    success=True,
                )

            if dg_new * (alpha_hi - alpha_lo) >= 0:
                alpha_hi = alpha_lo
                f_hi = f_lo

            alpha_lo = alpha
            f_lo = f_new

        # Bracket too small
        if abs(alpha_hi - alpha_lo) < 1e-14:
            break

    # Return best we found (alpha_lo is usually the better end)
    x_final = add_scaled(x, d, alpha_lo)
    return LineSearchResult(
        alpha=alpha_lo,
        f_new=f(x_final),
        g_new=grad(x_final),
        function_calls=function_calls + 1,
        gradient_calls=gradient_calls + 1,
        success=False,
    )
